### PR TITLE
feat!: Image Embeddings

### PR DESCRIPTION
**Pull request by lywa1998**
This PR includes the following major changes:  
1. Support ImageEmbedding，[issue 71](https://github.com/Anush008/fastembed-rs/issues/71)
2. Deleted unnecessary integration test directory

### CODE REVIEW

**Comment by Anush008** [comment]
There are breaking changes, which should be fine since we'll do a major release(v4) along with #99.

**Comment by lywa1998** [comment]
Thank you for your review; I have added content on how to use ImageEmbedding in the `README.md`.

File: README.md [hunk 1]
```diff
@@ -100,6 +107,30 @@ let documents = vec![
...
+    ..Default::default()
+})?;
+
+let images = vec!["assets/image_0.png", "assets/image_1.png"];
+
+// Generate embeddings with the default batch size, 256
+let embeddings = model.embed(images, None)?;
+
+println!("Embeddings length: {}", embeddings.len()); // -> Embeddings length: 4
+println!("Embedding dimension: {}", embeddings[0].len()); // -> Embedding dimension: 384
```
Comment (Anush008) [inline README.md:131]:
```suggestion
println!("Embeddings length: {}", embeddings.len()); // -> Embeddings length: 2
println!("Embedding dimension: {}", embeddings[0].len()); // -> Embedding dimension: 512
```

File: src/image_embedding.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+use rayon::{iter::ParallelIterator, slice::ParallelSlice};
+use std::{
+    fmt::Display,
+    path::{Path, PathBuf},
+    thread::available_parallelism,
+};
+const DEFAULT_BATCH_SIZE: usize = 256;
+const DEFAULT_EMBEDDING_MODEL: ImageEmbeddingModel = ImageEmbeddingModel::ClipVitB32;
+
+/// Options for initializing the TextEmbedding model
```
Comment (Anush008) [inline src/image_embedding.rs:22]:
```suggestion
/// Options for initializing the ImageEmbedding model
```

File: src/image_embedding.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+                let input_name = self.session.inputs[0].name.clone();
+                let session_inputs = ort::inputs![
+                    input_name => Value::from_array(pixel_values_array)?,
+                ]?;
+
+                let outputs = self.session.run(session_inputs)?;
+
+                // Try to get the only output key
+                // If multiple, then default to `last_hidden_state`
```
Comment (Anush008) [inline src/image_embedding.rs:237]:
```suggestion
                // If multiple, then default to `image_embeds`
```

File: src/models/image_embedding.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            dim: 512,
+            description: String::from("Unicom Unicom-ViT-B-32 from open-metric-learning"),
+            model_code: String::from("Qdrant/Unicom-ViT-B-32"),
+            model_file: String::from("model.onnx"),
+        }
+    ];
+
+    // TODO: Use when out in stable
+    // assert_eq!(
+    //     std::mem::variant_count::<EmbeddingModel>(),
```
Comment (Anush008) [inline src/models/image_embedding.rs:49]:
```suggestion
    //     std::mem::variant_count::<ImageEmbeddingModel>(),
```

File: README.md [hunk 5]
```diff
@@ -46,6 +46,13 @@ The default model is Flag Embedding, which is top of the [MTEB](https://huggingf
 
 - [**prithivida/Splade_PP_en_v1**](https://huggingface.co/prithivida/Splade_PP_en_v1) - Default
 
+### Image Embedding
```
Comment (Anush008) [inline README.md:49]: Not that `TokenizerFiles` is removed, this [README.md section](https://github.com/Anush008/fastembed-rs#:~:text=results)%3B-,Alternatively%2C%20raw,-.onnx%20files%20can) needs to be updated since it's much simpler now.

File: src/common.rs [hunk 6]
```diff
@@ -149,12 +42,305 @@ pub fn read_file_to_bytes(file: &PathBuf) -> Result<Vec<u8>> {
...
+
+impl Transform for Rescale {
+    fn transform(&self, data: TransformData) -> Result<TransformData> {
+        let array = data.array()?;
+        let array = array * self.scale;
+        Ok(TransformData::NdArray(array))
+    }
+}
+
+pub struct Normlize {
```
Comment (Anush008) [inline src/common.rs:172]: Typo? `Normalize`?

File: src/common.rs [hunk 7]
```diff
@@ -149,12 +42,305 @@ pub fn read_file_to_bytes(file: &PathBuf) -> Result<Vec<u8>> {
...
+        let shape = array.shape().to_vec();
+        match shape.as_slice() {
+            [c, h, w] => {
+                let array_normalized = array
+                    .sub(mean.broadcast((*c, *h, *w)).unwrap())
+                    .div(std.broadcast((*c, *h, *w)).unwrap());
+                Ok(TransformData::NdArray(array_normalized))
+            }
+            _ => Err(anyhow!(
+                "Transformer convert error. Normlize operator get error shape."
```
Comment (Anush008) [inline src/common.rs:196]: Typo? `Normalize`?

File: src/common.rs [hunk 8]
```diff
@@ -149,12 +42,305 @@ pub fn read_file_to_bytes(file: &PathBuf) -> Result<Vec<u8>> {
...
+            .map(|value| {
+                value
+                    .as_f64()
+                    .map(|num| num as f32)
+                    .ok_or(anyhow!("image_mean must be float"))
+            })
+            .collect::<Result<Vec<f32>>>()?;
+        let std = config["image_std"]
+            .as_array()
+            .ok_or(anyhow!("image_std must be contain"))?
```
Comment (Anush008) [inline src/common.rs:333]:
```suggestion
            .ok_or(anyhow!("image_std must be contained"))?
```

File: tests/jinaai_multiligual.rs [hunk 9]
```diff
@@ -1,42 +0,0 @@
-#![cfg(feature = "online")]
-
```
Comment (Anush008) [inline tests/jinaai_multiligual.rs:2]: 👍

**Review by Anush008 — COMMENTED** [review]
Thanks again @lywa1998. This was great to use.
I only have a few nits for you to consider.

File: README.md [hunk 10]
```diff
@@ -125,19 +156,17 @@ let results = model.rerank("what is panda?", documents, true, None);
...
 ```
 
-Alternatively, raw `.onnx` files can be loaded through the `UserDefinedEmbeddingModel` struct (for "bring your own" text embedding models) using `TextEmbedding::try_new_from_user_defined(...)`. Similarly, "bring your own" reranking models can be loaded using the `UserDefinedRerankingModel` struct and `TextRerank::try_new_from_user_defined(...)`. For example:
+Alternatively, raw `.onnx` files can be loaded through the `UserDefinedEmbeddingModel` struct (for "bring your own" text embedding models) using `TextEmbedding::try_new_from_user_defined(...)`. Similarly,
+"bring your own" image embedding models can be loaded useing the `UserDefinedImageEmbeddingModel` struct and `ImageEmbedding::try_new_from_user_defined(...)`,
+"bring your own" reranking models can be loaded using the `UserDefinedRerankingModel` struct and `TextRerank::try_new_from_user_defined(...)`.
+For example:
 
 ```rust
 macro_rules! local_model {
```
Comment (Anush008) [inline README.md:165]: Maybe remove the macro? Since it's simple now.

File: README.md [hunk 11]
```diff
@@ -125,19 +156,17 @@ let results = model.rerank("what is panda?", documents, true, None);
 println!("Rerank result: {:?}", results);
 ```
 
-Alternatively, raw `.onnx` files can be loaded through the `UserDefinedEmbeddingModel` struct (for "bring your own" text embedding models) using `TextEmbedding::try_new_from_user_defined(...)`. Similarly, "bring your own" reranking models can be loaded using the `UserDefinedRerankingModel` struct and `TextRerank::try_new_from_user_defined(...)`. For example:
+Alternatively, raw `.onnx` files can be loaded through the `UserDefinedEmbeddingModel` struct (for "bring your own" text embedding models) using `TextEmbedding::try_new_from_user_defined(...)`. Similarly,
+"bring your own" image embedding models can be loaded useing the `UserDefinedImageEmbeddingModel` struct and `ImageEmbedding::try_new_from_user_defined(...)`,
```
Comment (Anush008) [inline README.md:160]:
```suggestion
"bring your own" image embedding models can be loaded using the `UserDefinedImageEmbeddingModel` struct and `ImageEmbedding::try_new_from_user_defined(...)`,
```

File: README.md (hunk 10 above)
Comment (lywa1998) [reply README.md:165]: The ·local_model· macro only provides a way to write it and is not included in the crate. I think it can be retained.

Comment (Anush008) [reply README.md:165]: Right. I meant to simplify the README.

File: README.md [hunk 12]
```diff
@@ -125,19 +156,17 @@ let results = model.rerank("what is panda?", documents, true, None);
 println!("Rerank result: {:?}", results);
 ```
 
-Alternatively, raw `.onnx` files can be loaded through the `UserDefinedEmbeddingModel` struct (for "bring your own" text embedding models) using `TextEmbedding::try_new_from_user_defined(...)`. Similarly, "bring your own" reranking models can be loaded using the `UserDefinedRerankingModel` struct and `TextRerank::try_new_from_user_defined(...)`. For example:
+Alternatively, raw `.onnx` files can be loaded through the `UserDefinedEmbeddingModel` struct (for "bring your own" text embedding models) using `TextEmbedding::try_new_from_user_defined(...)`. Similarly,
+"bring your own" image embedding models can be loaded using the `UserDefinedImageEmbeddingModel` struct and `ImageEmbedding::try_new_from_user_defined(...)`,
+"bring your own" reranking models can be loaded using the `UserDefinedRerankingModel` struct and `TextRerank::try_new_from_user_defined(...)`.
```
Comment (Anush008) [inline README.md:161]:
```suggestion
Alternatively, local model files can be used for inference via the `try_new_from_user_defined(...)` methods of respective structs.
```

File: README.md (hunk 10 above)
Comment (Anush008) [reply README.md:163]: cc @lywa1998

**Comment by Anush008** [comment]
I'll have to see what's up with the CI.

**Comment by Anush008** [comment]
I see there's 

```console
test tests::test_rerank ... FAILED
test tests::test_user_defined_embedding_model ... ok
test tests::test_sparse_embeddings ... ok
test tests::test_user_defined_reranking_model ... FAILED
```

File: README.md (hunk 10 above)
Comment (Anush008) [reply README.md]: [`95ef48e` (#101)](https://github.com/Anush008/fastembed-rs/pull/101/commits/95ef48ec5f40e27fc7744ad642e19da68a186542)

**Comment by Anush008** [comment]
Hey @lywa1998. I've commited [`b43b53f` (#101)](https://github.com/Anush008/fastembed-rs/pull/101/commits/b43b53f1c225dfb199d815e9499ab81b4b9eebe4) to fix the failing reranker.

**Comment by lywa1998** [comment]
Thank @Anush008 for fixing the reranker.

**Comment by Anush008** [comment]
GitHub Actions 😫

**Comment by Anush008** [comment]
Looks like [`b43b53f` (#101)](https://github.com/Anush008/fastembed-rs/pull/101/commits/b43b53f1c225dfb199d815e9499ab81b4b9eebe4) should be applied to the text embedding implementation too.

**Comment by lywa1998** [comment]
Yes, I have modified it and am currently testing it locally. However, there are too many models 😂.

**Comment by Anush008** [comment]
I recommend you push the commit. Let's just wait for the CI to do it.

**Comment by Anush008** [comment]
Hey. Can you please resolve the conflicts? Can't run the tests without it.

**Comment by Anush008** [comment]
I guess, this does change the tokenizer.

@lywa1998,  how about we keep this PR to the image models only and revert

> Removed Tokenizer, https://github.com/Anush008/fastembed-rs/issues/97

?

**Comment by Anush008** [comment]
We try reducing `1e-6`?

**Comment by Anush008** [comment]
Also, FastEmbed uses
https://huggingface.co/Qdrant/bge-base-en-v1.5-onnx-Q

https://github.com/qdrant/fastembed/blob/49762a6d199b1c45c02bc661074c8e026a0c5011/fastembed/text/onnx_embedding.py#L29

**Comment by triandco** [comment]
@lywa1998, I haven't got a chance to clone your branch to test, but I can see the change that cause the test to fail is in ```src/text_embedding.rs```: the ```common::load_tokenizer``` was replaced with ```Tokenizer::from_bytes(model.tokenizer_file)```.  ```common::load_tokenizer``` explicitly call ```.with_truncation``` and ``` ```.add_special_tokens``` when creating tokenizers. This will affect the number of token output as well as its head and tails. The number of tokens changes the shape attention mask. 

The error in github pipeline is about the attention mask couldn't be expanded correctly for pooling:
```
thread 'tests::test_bgesmallen1point5_match_python_counterpart' panicked at src/text_embedding.rs:321:77:
called `Result::unwrap()` on an `Err` value: Failed to run inference on model: Non-zero status code returned while running Add node. Name:'/embeddings/Add_1' Status Message: /home/runner/work/ort-artifacts-staging/ort-artifacts-staging/onnxruntime/onnxruntime/core/providers/cpu/math/element_wise_ops.h:560 void onnxruntime::BroadcastIterator::Append(ptrdiff_t, ptrdiff_t) axis == 1 || axis == largest was false. Attempting to broadcast an axis by a dimension other than 1. 512 by 837
```

I recommend writing a little unit test to compare the size to the token output between tokenizer created with ```common::load_tokenizer``` and the plainer new version. Hope this helps. ^^

**Comment by lywa1998** [comment]
Thank @trindco for your participation. The main issue with GitHub Action is related to the export and use of `common::load_tokenizer` in `Xenova/bge-small-en-v1.5`. 

1. The model exported from `Xenova/bge-small-en-v1.5` limits the input dimensions to 512. As shown in the screenshot, the data in the Slice node is enforced to initialize as 1*512, which causes the model to throw an error if the input exceeds 512 tokens.
  
2. I tested the `comm::load_tokenizer` method in the `main` branch and found that the test `test::test_bgesmallen1point5_match_python_counterpart` passing is problematic. By examining the encoding `src/text_embedding:253`, it appears that the tokenizer only processed part of the sample text, particularly the item [2] in the sample text: "...17% of global greenhouse gas emissions." This indicates that the result of `common::load_tokenizer` is incorrect.

**Comment by triandco** [comment]
@lywa1998, I've added a pull request which details how i generate the base line embedding including the tokenizer settings i used.

Below is the python code for it, please excuse the messy copy paste code because I have this in a private repo. In this code, I run the onnx model with onnxruntime but i pre-process and post-process with a SentenceTransformer base to minimize introducing any new behaviour.

It's rather rudimentary and inefficient, but the output seems reasonable when comparing to the original torch model. I manually downloaded the [xenova base model from huggingface](https://huggingface.co/Xenova/bge-small-en-v1.5/tree/main/onnx). If you need to update the test or change settings you can alter the code to get a new baseline.

Much appreciate your work, image embedding would be a great addition.

```python
import json
import os
import torch
import numpy
from typing import Any, Dict, List, Protocol, Sequence, Union, runtime_checkable

from sentence_transformers import SentenceTransformer
import onnx
from onnxruntime import InferenceSession

class OnnxModel(Encoder):
    onnx_model: onnx.ModelProto
    torch_model: SentenceTransformer
    session: InferenceSession
    tokenizer: Any

    @staticmethod
    def create():
        return OnnxModel("_")

    def __init__(self, device: str | None = None) -> None:
        current_path = os.path.dirname(os.path.realpath(__file__))
        onnx_path = current_path + "/xenova-bge-small-en-v1point5.onnx"
        self.onnx_model = onnx.load(onnx_path)
        self.session = InferenceSession(onnx_path)
        self.torch_model = SentenceTransformer("BAAI/bge-small-en-v1.5")
        self.tokenizer = self.torch_model.tokenizer

    def encode_one(self, text: str):
        tokens = self.tokenizer(
            text, 
            padding="max_length",
            max_length=384,
            truncation=True,
            return_tensors="np")
        
        input = {
            'input_ids': tokens['input_ids'],
            'attention_mask': tokens['attention_mask'],
            # 'token_type_ids': tokens['token_type_ids'] -- this is necessary with qdrant onnx model
            }
        print(f"input_ids: {tokens['input_ids']}")
        print(f"attention mask: {tokens['attention_mask']}")
        # These outputs are the hidden state before pooling
        outputs = self.session.run(None, input)

        normalization = None
        for modules in self.torch_model.modules():
            for idx, module in enumerate(modules):
                if idx == 1:
                    pooling = module
                if idx == 2:
                    normalization = module
            break

        pooled = pooling.forward(
            features={
                "token_embeddings": torch.Tensor(outputs[0]),
                "attention_mask": torch.Tensor(input['attention_mask']),
            },
        )
        normalised = normalization.forward(features=pooled) if normalization is not None else pooled
        return normalised["sentence_embedding"][0].numpy()
    
    def encode(
        self, sentences: Sequence[str], *, prompt_name: str | None = None, **kwargs: Any
    ) -> torch.Tensor | numpy.ndarray:
        return numpy.asarray([ self.encode_one(text) for text in sentences ])
```

**Review by Anush008 — APPROVED** [review]
Following #109 and this, I guess that would be all for v4.

Thank you @lywa1998.
